vllm/model_executor/models/ultravox.py

```diff
@@ -423,9 +482,26 @@ def forward(self, input_ids: torch.Tensor, positions: torch.Tensor,
             inputs_embeds = self.language_model.model.get_input_embeddings(
                 input_ids)
 
-            inputs_embeds = merge_multimodal_embeddings(
-                input_ids, inputs_embeds, audio_embeddings,
-                _AUDIO_PLACEHOLDER_TOKEN)
+            prefill_metadata = attn_metadata.prefill_metadata
+            if not prefill_metadata:
+                raise ValueError(
+                    "Audio inputs should only be handled during prefill.")
+
+            seq_lens_tensor = getattr(prefill_metadata, "seq_lens_tensor",
+                                      None)
+            if seq_lens_tensor is None:
+                raise ValueError(
+                    "AttentionMetadata should include a seq_lens_tensor.")
+
+            seq_start_loc = getattr(prefill_metadata, "seq_start_loc", None)
+            if seq_start_loc is None:
+                raise ValueError(
+                    "AttentionMetadata should include seq_start_loc.")
```

Comment:
@mgoin is this true for all `AttentionMetadata`?

Reply:
Yeah this was the main thing I was uncertain about -- if there's a canonical "right" way to do this that is independent of model/backend/whatever I'd be inclined to push the `AttentionMetadata` down into the `merge_*` function and do it there.

Reply:
@comaniac maybe you can help answer this question as well.

Reply:
No it's not, but in this case you should directly check the attention type and error out otherwise.

Reply:
I noticed that several of the `AttentionMetadata` subtypes had these fields -- a couple questions:
- Do the ones that _don't_ have them support chunked prefill/prefix caching? If so is there a way to relate the input_ids/positions to the original sequences?
- What determines which `AttentionMetadata` is used? Model? Hardware? Options/CLI args?

Reply:
I pushed these conditions down into the merge embeddings helper and (somewhat indiscriminately) added an `isinstance` check for the various `AttentionMetadata` types that had these fields.

Reply:
Now each of the runners/attention backends computes the mapping explicitly.